google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/test/java/com/google/cloud/storage/contrib/nio/it/ITGcsNio.java

```diff
@@ -123,10 +136,162 @@ public static void afterClass() throws ExecutionException, InterruptedException
+    }
+  }
+
+  @Test
+  public void testCanCreateWithUserProject() throws IOException {
+    CloudStorageFileSystem testBucket = getRequesterPaysBucket(false, PROJECT);
+    Path path = testBucket.getPath(TMP_FILE);
+    // should succeed because we specified a project
+    Files.write(path, "I would like to write, please?".getBytes());
+  }
+
+  @Test
+  public void testCantReadWithoutUserProject() throws IOException {
+    CloudStorageFileSystem testBucket = getRequesterPaysBucket(false, "");
+    Path path = testBucket.getPath(SML_FILE);
+    try {
+      // fails
+      Files.readAllBytes(path);
+      Assert.fail("It should have thrown an exception.");
+    } catch (StorageException sex) {
+      assertIsRequesterPaysException(sex);
+    }
   }
 
+  @Test
+  public void testCanReadWithUserProject() throws IOException {
+    CloudStorageFileSystem testBucket = getRequesterPaysBucket(false, PROJECT);
+    Path path = testBucket.getPath(SML_FILE);
+    // should succeed because we specified a project
+    Files.readAllBytes(path);
+  }
+
+  @Test
+  public void testCantCopyWithoutUserProject() throws IOException {
+    CloudStorageFileSystem testRPBucket = getRequesterPaysBucket(false, "");
+    CloudStorageFileSystem testBucket = getTestBucket();
+    CloudStoragePath[] sources = new CloudStoragePath[] { testBucket.getPath(SML_FILE), testRPBucket.getPath(SML_FILE) };
+    CloudStoragePath[] dests = new CloudStoragePath[] {testBucket.getPath(TMP_FILE), testRPBucket.getPath(TMP_FILE) };
+    for (int s=0; s<2; s++) {
+      for (int d=0; d<2; d++) {
+        // normal to normal is out of scope of RP testing.
+        if (s==0 && d==0) {
+          continue;
+        }
+        String sdesc = (s==0?"normal bucket":"requester-pays bucket");
+        String ddesc = (d==0?"normal bucket":"requester-pays bucket");
+        try {
+          System.out.println("Copying from " + sdesc + " to " + ddesc);
+          Files.copy(sources[s], dests[d]);
+          Assert.fail("Shouldn't have been able to copy from " + sdesc + " to " + ddesc);
+          // for some reason this throws "GoogleJsonResponseException" instead of "StorageException"
+          // when going from requester pays bucket to requester pays bucket, but otherwise we get a
+          // normal StorageException.
+        } catch (HttpResponseException hex) {
+          Assert.assertEquals(hex.getStatusCode(), 400);
+          Assert.assertTrue(hex.getMessage().contains("Bucket is requester pays bucket but no user project provided"));
+        } catch (StorageException sex) {
+          assertIsRequesterPaysException(sex);
+        }
+      }
+    }
+  }
+
+  @Test
+  public void testCanCopyWithUserProject() throws IOException {
+    CloudStorageFileSystem testRPBucket = getRequesterPaysBucket(false, PROJECT);
+    CloudStorageFileSystem testBucket = getTestBucket();
+    CloudStoragePath[] sources = new CloudStoragePath[] { testBucket.getPath(SML_FILE), testRPBucket.getPath(SML_FILE) };
```

Comment:
Neat: please remove unnecessary white spaces

Reply:
And the one before `}` as well :)

Reply:
Ah yes, I had removed that one too, just forgot to mention it here. All good then!